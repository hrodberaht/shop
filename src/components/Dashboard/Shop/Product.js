import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class Product extends Component {
  state = {
    pcsOrder: 0,
    error: null,
  };

  changeInStockToText = (inStock) => {
    if (inStock > 100) return 'full supply';
    if (inStock > 10 && inStock <= 100) return 'medium supply';
    if (inStock > 0 && inStock <= 10) return 'last pieces';

    return 'not available';
  };

  handleChange = (e) => {
    const {
      product: { inStock },
    } = this.props;
    this.setState({ pcsOrder: e.target.value });
    this.checkAmountOrder(+inStock, +e.target.value);
  };

  checkAmountOrder = (inStock, pcsOrder) => {
    if (pcsOrder > inStock) return this.setState({ error: 'Not enough products' });
    if (pcsOrder < 0) return this.setState({ pcsOrder: 0 });
    return this.setState({ error: null });
  };

  calculateTotalPrice = (pcsOrder, price) => pcsOrder * price;

  render() {
    const {
      product: {
        name, type, price, inStock,
      },
    } = this.props;
    const { pcsOrder, error } = this.state;
    return (
      <div className="product-item">
        <h3>{name}</h3>
        <h5>{type}</h5>
        <p className="product-item__price">
          <span>$</span>
          {price}
        </p>
        <p>
          <span>In stock:</span>
          {'  '}
          <span id="stock-message">{this.changeInStockToText(+inStock)}</span>
        </p>
        <p className="product-item__order">
          {' '}
          Order:
          <input type="number" value={pcsOrder} onChange={this.handleChange} />
          pcs
        </p>
        <p id="stock-error">{error}</p>
        <p className="product-item__total">
          <span>Total price:</span>
          {'  '}
          <span id="total-price">
            <span>$</span>
            {this.calculateTotalPrice(pcsOrder, price)}
          </span>
        </p>
        <button className="btn btn-primary" type="button">
          Buy
        </button>
      </div>
    );
  }
}

Product.propTypes = {
  product: PropTypes.shape({
    name: PropTypes.string,
    type: PropTypes.string,
    price: PropTypes.string,
    inStock: PropTypes.string,
  }).isRequired,
};
