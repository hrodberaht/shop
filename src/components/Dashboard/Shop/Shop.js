import React, { Component } from 'react';
import ConnectedProductsList from './ProductsList';

export default class Shop extends Component {
  render() {
    return (
      <div className="shop">
        <ConnectedProductsList />
      </div>
    );
  }
}
