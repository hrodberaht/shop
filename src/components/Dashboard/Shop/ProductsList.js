import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Product from './Product';
import { getProducts } from '../../../store/products/selectors';

export class ProductsList extends Component {
  render() {
    const { products } = this.props;
    return (
      <div>
        {products.map(product => (
          <Product key={product.id} product={product} />
        ))}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  products: getProducts(state),
});

export default connect(mapStateToProps)(ProductsList);

ProductsList.propTypes = {
  products: PropTypes.arrayOf(PropTypes.object).isRequired,
};
