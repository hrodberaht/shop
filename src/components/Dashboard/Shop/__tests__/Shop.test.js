import Shop from '../Shop';

const setup = buildSetup(Shop, {});

describe('<Shop />', () => {
  it('should render without crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
