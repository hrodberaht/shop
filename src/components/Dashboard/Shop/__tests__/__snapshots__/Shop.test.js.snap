// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Shop /> should render without crash 1`] = `
<div
  className="shop"
>
  <Connect(ProductsList) />
</div>
`;
